tests/integration/security/pass_through_filter_chain_test.go

```diff
@@ -168,10 +168,15 @@ func TestPassThroughFilterChain(t *testing.T) {
 							if response.StatusCodeOK != responses[0].Code {
 								return fmt.Errorf("want status %s but got %s", response.StatusCodeOK, responses[0].Code)
 							}
-						} else if err == nil || !strings.Contains(err.Error(), "EOF") {
-							// The request should always be rejected in TCP level because we currently do not support
-							// HTTP filters in pass through filter chains.
-							return fmt.Errorf("want error EOF but got: %v", err)
+						} else {
```

Comment:
I think with the inbound HTTP protocol sniffing, all the request sent from the test app is probably treated as HTTP which means we are not testing the TCP filter chain in this case, could you update the test cases to make sure we have coverage for the TCP filter chain? (you might need some updates to the test framework, I think it always use HTTP underline, cc @nmittler )

Reply:
I think the issue is here for a long time. The test framework doesn't support test TCP traffic other than using HTTP.

Reply:
I think adding it is a blocker for merging this, as is as we are effectively losing coverage on critical RBAC tests if we merge right now

Reply:
I am happy to see someone pick the task and add support in test framework to support TCP traffic. Adding the test coverage for TCP filter chain is easy work.

Reply:
The code is under `pkg/test/echo`, it shouldn't be too much work for you too add

Reply:
It may take me some time to work that as I am not familiar with that code. I prefer to duplicate the tests with another one to disable sniffing which tests the TCP filter chain. Even though the test doesn't cover the RBAC for TCP right now, it covers the RBAC for HTTP which is not covered in previous scenario.

Reply:
That doesn't test this feature though. Even if you do that, for all we know in the default case with protocol sniffing enabled, RBAC is bypassed.

Reply:
RBAC is not bypassed when protocol sniffing is enabled. The test is for testing RBCA applied to pass through filter chain. If the RBAC is applied to pass through filter chain, it cannot be bypassed. The test has covered this case.

Reply:
> Even though the test doesn't cover the RBAC for TCP right now

This is incorrect, the current test covers RBAC for TCP because the inbound protocol sniffing is disabled and the pass through filter chain only has one TCP filter chain with the RBAC filter config. 

It becomes uncovered only after enabled the inbound protocol sniffing in this PR because an extra pass through filter chain for HTTP is added, and the new HTTP filter chain effectively bypass the TCP pass through filter chain.

Reply:
The test case cover some cases which applies authz policy and the server returns 403 by RBCA filter

Reply:
>  the new HTTP filter chain effectively bypass the TCP pass through filter chain

is this bypass? since both filter chain have the same RBAC rules

Reply:
You do not cover the case where sniffing is enabled and we send TCP traffic, as this cannot be done without adding support

Reply:
> This is incorrect, the current test covers RBAC for TCP because the inbound protocol sniffing is disabled and the pass through filter chain only has one TCP filter chain with the RBAC filter config.

I mean after the fix. The test case covers HTTP but doesn't cover TCP as all traffic is inspected as HTTP

Reply:
> is this bypass? since both filter chain have the same RBAC rules

I don't think we are saying that the code is broken, just that there is a testing gap

Reply:
> The test case covers HTTP but doesn't cover TCP as all traffic is inspected as HTTP

This is exactly what I am saying I am not comfortable merging until this is fixed so the tests cover this case.

Reply:
> You do not cover the case where sniffing is enabled and we send TCP traffic, as this cannot be done without adding support

That's why I said we can duplicate the tests. One enables sniffing, one disable

Reply:
Even if you duplicate the test... you are still not covering the case where sniffing is enabled AND you send TCP traffic. You can only cover the case where you send HTTP traffic

Reply:
By duplicating the test, both HTTP and TCP filter chain will be tested

Reply:
How can you test the TCP filter chain without sending TCP requests? Every request will match the HTTP filter?

Reply:
If we disable the sniffing, sending HTTP requests will test the TCP filter chain. disabling the sniffing + sending HTTP request ~ enabling the sniffing + sending TCP request

Reply:
I am looking for a test case with the following conditions:
* Protocol sniffing is enabled
* We send TCP traffic

This cannot be done without changes to the echo code. The two tests you mentioned do not satisfy these

Reply:
The test framework doesn't support protocol sniffing is enabled + TCP traffic. But protocol sniffing is enabled + TCP traffic equals to protocol sniffing is disabled + HTTP traffic. 

Protocol sniffing is enabled + TCP traffic -> TCP catch all filter chain
Protocol sniffing is disabled + HTTP traffic -> TCP catch all filter chain

Reply:
> But protocol sniffing is enabled + TCP traffic equals to protocol sniffing is disabled + HTTP traffic.

In theory... the only way you will convince me is if you have a test that **proves** this is true though. I am not trying to be annoying here, but this is RBAC we are talking about. We cannot risk a bypass. Just because the code is meant to work a certain way doesn't mean its ok to not test it, especially to make sure we don't have regression.

> The test framework doesn't support protocol sniffing is enabled + TCP traffic

As the person adding protocol sniffing, I think the burden is on you to make sure your feature is tested. If this means extending the test framework to handle this case, then the test framework needs to be extended. We cannot introduce risk just because the test framework doesn't support something

Reply:
Let me try to add the tcp for test framework